events/api/src/main/java/org/projectnessie/events/api/ContentKey.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (C) 2023 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.events.api;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import org.immutables.value.Value;
+
+/** Key for a {@link Content} object. */
+@Value.Immutable
+@JsonSerialize
+@JsonDeserialize(as = ImmutableContentKey.class)
+public interface ContentKey {
+
+  static ContentKey of(String... elements) {
+    return ImmutableContentKey.of(Arrays.asList(elements));
+  }
+
+  static ContentKey of(Iterable<String> elements) {
+    return ImmutableContentKey.of(elements);
+  }
+
+  /** The elements of the key. */
+  @Value.Parameter
+  List<String> getElements();
+
+  /**
+   * The name of the key.
+   *
+   * <p>The key name is the last element of the key. For example, the name of the key {@code ["a",
+   * "b", "c"]} is {@code "c"}.
+   */
+  @Value.Lazy
+  @JsonIgnore
+  default String getName() {
```

Comment:
Should we add a getter for the encoded full name string (as we have in `Namespace`)?

Reply:
I gave this a try in d0f908b, let me know what you think.

Reply:
Wonder whether it makes sense to introduce a new module that's used by `:nessie-model` and `:nessie-events-api` for all the relevant conversions, so we don't have to maintain those in multiple places. WDYT? Or even have a `BaseContentKey` class used by both?

Reply:
Maybe also for the actual `Content` and derived types. If, then it's probably worth to do it in a separate PR?

Reply:
+1 to extracting common code into a separate module.

Reply:
Validation and such can be specific to the "concrete" `Content` (et al) types in `:nessie-model` and `:nessie-event-api`. I was (maybe naively) thinking of a common abstract base class or interface in a new module and have concrete `@Immutable`s with the "right annotations".

Reply:
Ah yes, that could work 👍